### ISSUE TITLE
unnecessary_cast should catch unnecessary pointer cast

### ISSUE THREAD

**Issue by antonilol** [rust-lang/rust-clippy#12738]
### Summary

unnecessary_cast should catch unnecessary pointer casts like `&u32` to `*const u32` when this would already be done implicitly

edge case to consider:
it is not always converted implicitly, like when changing the type like this: `a as *const u32 as *const i32`, with `a: &u32`

### Lint Name

unnecessary_cast

### Reproducer

I tried this code:

```rust
fn main() {
    let a = 123u32;
    let _: u32 = a as u32;

    let a = &123u32;
    let _: *const u32 = a as *const u32;
}
```

I expected to see this happen:
get a warning from the lint for both casts

Instead, this happened:
there only is a warning for the u32 -> u32 cast

### Version

```text
rustc 1.77.0 (aedd173a2 2024-03-17)
binary: rustc
commit-hash: aedd173a2c086e558c2b66d3743b344f977621a7
commit-date: 2024-03-17
host: x86_64-unknown-linux-gnu
release: 1.77.0
LLVM version: 17.0.6
```
